src/Middleware/RequestLimiter/src/RequestLimiterMiddleware.cs

```diff
@@ -0,0 +1,118 @@
+        {
+            _next = next;
+            _logger = loggerFactory.CreateLogger<RequestLimiterMiddleware>();
+            _options = options.Value;
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            _logger.LogInformation("Resource limiting: " + context.Request.Path);
+
+            var endpoint = context.GetEndpoint();
+            var attributes = endpoint?.Metadata.GetOrderedMetadata<RequestLimitAttribute>();
+
+            if (attributes == null)
+            {
+                await _next.Invoke(context);
+                return;
+            }
+
+            var resourceLeases = new Stack<ResourceLease>();
+            try
+            {
+                foreach (var attribute in attributes)
+                {
+                    if (!string.IsNullOrEmpty(attribute.Policy) && attribute.Limiter != null)
+                    {
+                        throw new InvalidOperationException("Cannot specify both policy and limiter");
+                    }
+
+                    if (string.IsNullOrEmpty(attribute.Policy) && attribute.Limiter == null)
+                    {
+                        if (_options.ResolveDefaultRequestLimit != null)
+                        {
+                            if (!await ApplyLimitAsync(_options.ResolveDefaultRequestLimit(context.RequestServices), context, resourceLeases))
+                            {
+                                return;
+                            }
+                        }
+                    }
+
+                    // Policy based limiters
+                    if (!string.IsNullOrEmpty(attribute.Policy))
+                    {
+                        if (!_options.PolicyMap.TryGetValue(attribute.Policy, out var policy))
+                        {
+                            throw new InvalidOperationException("Policy not found");
+                        }
+
+                        foreach (var limitResolver in policy.LimiterResolvers)
+                        {
+                            if (!await ApplyLimitAsync(limitResolver(context.RequestServices), context, resourceLeases))
+                            {
+                                return;
+                            }
+                        }
+                    }
+
+                    if (attribute.Limiter != null)
+                    {
+                        // Registrations based limiters
+                        if (!await ApplyLimitAsync(attribute.Limiter, context, resourceLeases))
+                        {
+                            return;
+                        }
+                    }
+
+                }
+
+                await _next.Invoke(context);
+            }
+            finally
+            {
+                while (resourceLeases.TryPop(out var resource))
+                {
+                    _logger.LogInformation("Releasing resource");
```

Comment:
Debug, and we want some way to identity the resource being released like the policy name.

Reply:
Ooops, this is meant to be placeholder logging.